File: sdks/nodejs/src/policy/enforcer.ts
import { CyberArmorClient } from '../client';
import { Decision, DecisionType, isAllowed, PolicyViolationError } from './decision';

export interface EvaluatePolicyOptions {
  action: string;
  provider?: string;
  model?: string;
  toolName?: string;
  promptText?: string;
  dataClassifications?: string[];
  humanInitiatorPresent?: boolean;
  environment?: string;
  sensitivityTier?: string;
}

export class PolicyEnforcer {
  constructor(private readonly ca: CyberArmorClient) {}

  async evaluate(options: EvaluatePolicyOptions): Promise<Decision> {
    const start = Date.now();
    try {
      const controller = new AbortController();
      const timeout = setTimeout(() => controller.abort(), this.ca.config.timeoutMs);

      const resp = await fetch(`${this.ca.config.controlPlaneUrl}/policies/evaluate`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'x-api-key': this.ca.config.agentSecret,
        },
        body: JSON.stringify({
          agent_id: this.ca.config.agentId,
          action_type: options.action,
          ai_provider: options.provider ?? 'unknown',
          model: options.model ?? 'unknown',
          tool_name: options.toolName,
          prompt_text: options.promptText?.slice(0, 2000),
          data_classifications: options.dataClassifications ?? [],
          human_initiator_present: options.humanInitiatorPresent ?? false,
          environment: options.environment ?? 'production',
          sensitivity_tier: options.sensitivityTier ?? 'internal',
        }),
        signal: controller.signal,
      });

      clearTimeout(timeout);

      if (resp.ok) {
        const data: any = await resp.json();
        return {
          decision: (data.decision as DecisionType) ?? DecisionType.ALLOW,
          reasonCode: data.reason_code ?? 'POLICY_PASSED',
          riskScore: data.risk_score ?? 0,
          policyId: data.policy_id,
          redactionTargets: data.redaction_targets ?? [],
          explanation: data.explanation ?? '',
          latencyMs: Date.now() - start,
        };
      }
    } catch {
      // Fallback to local evaluation
    }

    return this._localFallback(options, Date.now() - start);
  }

  enforce(decision: Decision): void {
    if (!isAllowed(decision)) {
      throw new PolicyViolationError(decision);
    }
  }

  private _localFallback(options: EvaluatePolicyOptions, latencyMs: number): Decision {
    if (this.ca.config.enforceMode === 'monitor') {
      return {
        decision: DecisionType.ALLOW_WITH_AUDIT_ONLY,
        reasonCode: 'MONITOR_MODE',
        riskScore: 0,
        latencyMs,
        explanation: 'Monitor mode: logging only',
      };
    }
    return {
      decision: DecisionType.ALLOW,
      reasonCode: 'LOCAL_FALLBACK_ALLOW',
      riskScore: 0,
      latencyMs,
      explanation: 'Control plane unreachable; local fallback',
    };
  }
}

File: sdks/go/policy/enforcer.go
package policy

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"time"
)

// EvaluatePolicyOptions holds the context for a policy evaluation request.
type EvaluatePolicyOptions struct {
	Action                string
	Provider              string
	Model                 string
	ToolName              string
	PromptText            string
	DataClassifications   []string
	HumanInitiatorPresent bool
	Environment           string
	SensitivityTier       string
}

// PolicyEnforcerConfig is the minimal interface the enforcer needs from the client.
type PolicyEnforcerConfig interface {
	GetControlPlaneURL() string
	GetAgentID() string
	GetAgentSecret() string
	GetEnforceMode() string
	GetHTTPClient() *http.Client
	IsFailOpen() bool
}

// Enforcer evaluates policies for AI agent actions.
type Enforcer struct {
	cfg PolicyEnforcerConfig
}

// NewEnforcer creates a new policy enforcer.
func NewEnforcer(cfg PolicyEnforcerConfig) *Enforcer {
	return &Enforcer{cfg: cfg}
}

// Evaluate evaluates a policy decision, falling back to local evaluation on CP failure.
func (e *Enforcer) Evaluate(ctx context.Context, opts EvaluatePolicyOptions) (Decision, error) {
	start := time.Now()
	decision, err := e.evaluateRemote(ctx, opts)
	if err != nil {
		return e.localFallback(opts, int(time.Since(start).Milliseconds())), nil
	}
	decision.LatencyMs = int(time.Since(start).Milliseconds())
	return decision, nil
}

// Enforce returns an error if the decision blocks the action.
func (e *Enforcer) Enforce(d Decision) error {
	if !d.IsAllowed() {
		return &PolicyViolationError{Decision: d}
	}
	return nil
}

func (e *Enforcer) evaluateRemote(ctx context.Context, opts EvaluatePolicyOptions) (Decision, error) {
	payload, _ := json.Marshal(map[string]any{
		"agent_id":                e.cfg.GetAgentID(),
		"action_type":             opts.Action,
		"ai_provider":             opts.Provider,
		"model":                   opts.Model,
		"tool_name":               opts.ToolName,
		"prompt_text":             truncate(opts.PromptText, 2000),
		"data_classifications":    opts.DataClassifications,
		"human_initiator_present": opts.HumanInitiatorPresent,
		"environment":             opts.Environment,
		"sensitivity_tier":        opts.SensitivityTier,
	})

	ctx2, cancel := context.WithTimeout(ctx, 3*time.Second)
	defer cancel()

	req, _ := http.NewRequestWithContext(ctx2, http.MethodPost,
		e.cfg.GetControlPlaneURL()+"/policies/evaluate",
		bytes.NewReader(payload))
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("x-api-key", e.cfg.GetAgentSecret())

	resp, err := e.cfg.GetHTTPClient().Do(req)
	if err != nil {
		return Decision{}, fmt.Errorf("policy remote eval: %w", err)
	}
	defer resp.Body.Close()

	var d Decision
	if err := json.NewDecoder(resp.Body).Decode(&d); err != nil {
		return Decision{}, fmt.Errorf("decode policy response: %w", err)
	}
	return d, nil
}

func (e *Enforcer) localFallback(opts EvaluatePolicyOptions, latencyMs int) Decision {
	if e.cfg.GetEnforceMode() == "monitor" {
		return Decision{Type: DecisionTypeAllowWithAuditOnly, ReasonCode: "MONITOR_MODE",
			Explanation: "Monitor mode — logging only", LatencyMs: latencyMs}
	}
	return Decision{Type: DecisionTypeAllow, ReasonCode: "LOCAL_FALLBACK_ALLOW",
		Explanation: "Control plane unreachable; local fallback", LatencyMs: latencyMs}
}

func truncate(s string, max int) string {
	if len(s) <= max {
		return s
	}
	return s[:max]
}
